js/pages/Route.js

```diff
@@ -11,7 +11,7 @@ define([
 		checkPermission() {
 			if (authApi.authProvider() === authApi.AUTH_PROVIDERS.IAP) {
 				return authApi.loadUserInfo();
-			} else if (appConfig.userAuthenticationEnabled && authApi.token() != null && authApi.tokenExpirationDate() > new Date()) {
+			} else if (appConfig.userAuthenticationEnabled && authApi.token() != null && this.timeToExpire() < appConfig.refreshTokenThreshold) {
```

Comment:
Permissions are updated along with the token refresh, so this logic will stop the permissions update until the threshold (which will be 8 hours from the moment of login with default settings: 12 hours token lifetime minus 4 hours threshold).

So if another user gave (or revoked) this user a pemission to some cohort/cs/etc., then the user will have to refresh the page to recieve updated permissions. In case of revoked permission, user will lose their changes and recieve 403 while trying to save the asset (and refresh the token and permissions list in the process, see [js/services/AuthAPI.js:171](https://github.com/OHDSI/Atlas/blob/master/js/services/AuthAPI.js#L171)).

Maybe consider refreshing permissions at least when open the asset editor?

Reply:
I believe that if you make a request that would violate permissions, a permission refresh is triggered (I believe it checks for status 403).  See [here](https://github.com/OHDSI/Atlas/blob/21c925c3b933929944ec0e0e8055db712214ebed/js/services/AuthAPI.js#L165).


My understanding/expectation is that when we get an exception that results in an accessDenied, it calls out to refrshToken(), which should call out to user/refresh, and then process the permissions.

The client state of permissions should have no impact on the permissions on the server.   if a permission is changed, it should be immediately in effect on the server, but potentially delayed from being known on the client.   If the code works as I think, then once we find out of a permission problem, we refresh perms from the server (and the UI will reflect the new permissions).

We could refresh permissions on every request, but that would bring us back to the original problem being solved here.

Reply:
I've added permission refreshes on load, create (because we need to know the new one has new permissions), and copy.   That should cover the cases where we have permissions changes during login.

Longer term, we need another way to check permissions on entities.  It should be a simple flag on an entity if you have read/write access to it (sources may work slightly differently), but the current approach means we download nearly a megabyte of permission strings and parse them every time we want to ask a simple question 'is this element readable/writeable'.